### DISCUSSION TITLE
Batch/Lot numbers full tracking support

### DISCUSSION THREAD

**Discussion by codebycab** [medusajs/medusa#3848] [Feature Requests]
There are some industries that **must** have full tracking of products by lot or batch numbers (see beauty, food, medicine). Searching for it in discussions or issue results in nothing.

Is this something difficult to implement? Is it in the roadmap?

**Comment by srindom** [comment]
Thanks for opening this - could you provide a more detailed description of what would be required?

**Comment by codebycab** [comment]
Sure,
A good feature to have alongside the inventory is a "lot number tracking", which means having an additional field when modifying the availability of products.

I'll give you an example:
Let's take a beauty product company, each product is part of a production lot.
If, after the sale, something goes wrong with the product, the company needs to know from which lot the defective product came from, in order to take the appropriate actions (recall each product from that specific lot, review manufacturing defects, etc...)

Now, this could be done using variants. Simply having 50 items of variant ABC001 (lot number) and 20 of variant ABC002. When the customer buys the product we automatically choose which variant (read: lot number) to sell them.

While this approach could work, it means that the variants feature wouldn't really be used for it's intended purpose.

**Reply by srindom** [reply]
Thanks for adding the extra details here very helpful! I will map out current inventory module setup and then provide a potential path to supporting something like this - let me know if you think this could work.

1. Currently, we have:
    - **ProductVariants**: represent purchasable entities
    - **InventoryItems**: represent stockable items
    - **InventoryLevels**: represent the quantity of a given item at a given location
    - ProductVariants are connected with one or more InventoryItems through a ProductVariantInventoryItem entity.
2. Let's say we have a Product that needs to be "batch controlled". The product has one ProductVariant, and there are two batches for the variant batch A and B.
3. Assumption: a consumer won't be able to decide if they purchase from batch A or batch B. Instead they simply purchase the ProductVariant and end up receiving from a (seemingly) arbitrary batch.
4. To support the two different batches we create two InventoryItems one for batch A and one for batch B. Both of the InventoryItems are associated with the ProductVariant. (I.e., we create two ProductVariantInventoryItem associations that both point to the same ProductVariant).
5. With this setup Medusa core needs to add logic along the lines of:
    - For purchases of the ProductVariant you can fulfill the order using stock from either InventoryItem for batch A or batch B.
    

Do you think this would be a good setup? 

--

Currently, we support adding multiple items to the same ProductVariant, but only to allow assembly products (e.g., a table consists of an InventoryItem for a table top and 4 inventory items for table legs). A way to think of this is that you can create AND conditions for ProductVariants and InventoryItems (fulfill ProductVariant using table top AND table legs). If we add support for OR behavior as well that should provide the necessary data model to support batch control (e.g., fulfill ProductVariant using batch A OR batch B).

**Comment by pevey** [comment]
I just want to add that I also have a business that has to track lot numbers.  This is a coffee roasting business, which is regulated by the FDA in the US and requires tracking in case there is a recall.  We currently track batch numbers by keeping a pen and paper "roast log" and then adding batch numbers from the roast log in our shipping metadata.

It would be awesome to have a way to be able to handle this in Medusa.  Far beyond expectations.  I know dispensaries in the US also have to deal with this tracking ("seed to sale") with even stricter oversight.  

The best way to go about may be to do with fulfilment rather than setting up the products.  Batches may be a concept tied to fulfilment in a similar way to variants being attached to products.  There could be any number of batches, and the expectation would be that more and more are added over time.  The records of where each batch is shipped only have to be kept for so long (can vary based on product), so a purge option would also be really nice.
